iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,429 @@
+///         http://eel.is/c++draft/views contains the latest C++20 draft
+/// @tparam T - element type; must be a complete object type that is not an abstract class type
+/// @tparam Extent - the number of elements in the sequence, or std::dynamic_extent if dynamic
+template <typename T, uint64_t Extent>
+class span : public detail::span_storage<Extent>
+{
+  private:
+    using span_storage_t = detail::span_storage<Extent>;
+
+  public:
+    using element_type = T;
+    using value_type = std::remove_cv_t<T>;
+    using uint64_type = uint64_t;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T*;
+    using reference = T&;
+    using iterator = span_iterator<T>;
+    using reverse_iterator = std::reverse_iterator<iterator>;
+    static constexpr uint64_t extent = Extent;
+
+    // constructors, copy, assignment, and destructor
+
+    constexpr span() noexcept = delete;
+
+    /// @brief Constructs a span that is a view over the range [first, first + count);
+    /// @tparam It Type of the iterator
+    /// @param first iterator of where to start creating the span
+    /// @param count number of elements counting from first
+    template <typename It>
+    constexpr span(It first, uint64_t count) noexcept;
+
+    /// @brief Constructs a span that is a view over the range [first, last);
+    /// @tparam It Type of the start iterator
+    /// @tparam End Type of the end iterator
+    /// @param begin iterator of where to start creating the span
+    /// @param end iterator of where to stop
+    template <typename It, typename End, typename = std::enable_if_t<!std::is_convertible<End, uint64_t>::value>>
+    constexpr span(It begin, End end) noexcept;
+
+    /// @brief Constructs a span that is a view over the array arr; the resulting span has size() == N and data() ==
+    /// std::data(arr)
+    /// @tparam N implicit size of the array
+    /// @param array used to construct the span
+    // NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+    template <uint64_t N, typename = detail::enable_if_compatible_array_t<T (&)[N], T, Extent>>
+    constexpr explicit span(T (&array)[N]) noexcept;
+    // NOLINTEND(cppcoreguidelines-avoid-c-arrays, hicpp-avoid-c-arrays)
+
+    /// @brief Constructs a span that is a view over the uninitialized array; the resulting span has size() == N and
+    /// data() == std::data(arr)
+    /// @tparam U implicit element type stored in the uninitialized array
+    /// @tparam N implicit capacity of the uninitialized array
+    /// @param array uninitialized array used to create the span
+    template <typename U,
+              uint64_t N,
+              template <typename, uint64_t>
+              class Buffer,
+              typename = detail::enable_if_compatible_array_t<iox::UninitializedArray<U, N, Buffer>&, T, Extent>>
+    constexpr explicit span(iox::UninitializedArray<U, N, Buffer>& array) noexcept;
```

Comment:
Just as food for thought to get rid of the dependency to `UninitializedArray`. One could have something like
```cpp
template <typename T>
SpanCompatibleContainerAdapter {
    static_assert(always_false_v<T>, "Please specialize for your container");
    
    // some methods which are called by the ctor
}

// in the span
template <typename Container, typename = detail::enable_if_compatible_dynamic_container_t<Container&, T, Extent>>
constexpr explicit span(SpanCompatibleContainerAdapter<Container>& container) noexcept;
```

The nice thing about this approach would be to be able to create new Container and make them useable with the `span` without having to adjust the `span` itself.

Disclainer: I did not yet look at the implementation so the above might not make sense or would need adjustments.

Reply:
@elBoberido That's a nice future improvement, thanks! Let's for now stick with the direct dependency.